src/diffusers/models/embeddings.py

```diff
@@ -2104,6 +2104,98 @@ def forward(self, id_embeds: torch.Tensor) -> torch.Tensor:
         return out
 
 
+class IPAdapterTimeImageProjectionBlock(nn.Module):
+    def __init__(
+        self,
+        hidden_dim: int = 768,
+        dim_head: int = 64,
+        heads: int = 16,
+        ffn_ratio: float = 4,
+    ) -> None:
+        super().__init__()
+        from .attention import FeedForward
+
+        self.ln0 = nn.LayerNorm(hidden_dim)
+        self.ln1 = nn.LayerNorm(hidden_dim)
+        self.attn = Attention(
+            query_dim=hidden_dim,
+            cross_attention_dim=hidden_dim,
+            dim_head=dim_head,
+            heads=heads,
+            bias=False,
+            out_bias=False,
+            processor=FusedAttnProcessor2_0(),
+        )
+        self.ff = FeedForward(hidden_dim, hidden_dim, activation_fn="gelu", mult=ffn_ratio, bias=False)
+
+        # AdaLayerNorm
+        self.adaln_silu = nn.SiLU()
+        self.adaln_proj = nn.Linear(hidden_dim, 4 * hidden_dim)
+        self.adaln_norm = nn.LayerNorm(hidden_dim)
+
+        # Custom scale cannot be passed in constructor
+        self.attn.scale = 1 / math.sqrt(math.sqrt(dim_head))
+        self.attn.fuse_projections()
+        self.attn.to_k = None
+        self.attn.to_v = None
+
+    def forward(self, x, latents, timestep_emb):
+        shift_msa, scale_msa, shift_mlp, scale_mlp = self.adaln_proj(self.adaln_silu(timestep_emb))
+
+        x = self.ln0(x)
+        latents = self.ln1(latents) * (1 + scale_msa[:, None]) + shift_msa[:, None]
+        latents = self.attn(x, latents) + latents
+
+        residual = latents
+        latents = self.adaln_norm(latents) * (1 + scale_mlp[:, None]) + shift_mlp[:, None]
+        return self.ff(latents) + residual
```

Comment:
Have you verified the numerical accuracy of this vs the original?

Reply:
I'm testing it right now, the former `TimePerceiverAttention` class can be replaced by `Attention` and `FusedAttnProcessor2_0` processor, but that would require doing some conversions in the state_dict. As I have it right now, we need to change these keys:

```python
image_proj_state_dict = {}
for key, value in state_dict["image_proj"].items():
    for idx in range(4):
        key = key.replace(f"layers.{idx}.0.norm1", f"layers.{idx}.ln0")
        key = key.replace(f"layers.{idx}.0.norm2", f"layers.{idx}.ln1")
        key = key.replace(f"layers.{idx}.0.to_q", f"layers.{idx}.attn.to_q")
        key = key.replace(f"layers.{idx}.0.to_kv", f"layers.{idx}.attn.to_kv")
        key = key.replace(f"layers.{idx}.0.to_out", f"layers.{idx}.attn.to_out.0")
        key = key.replace(f"layers.{idx}.1.0", f"layers.{idx}.adaln_norm")
        key = key.replace(f"layers.{idx}.1.1", f"layers.{idx}.ff.net.0.proj")
        key = key.replace(f"layers.{idx}.1.3", f"layers.{idx}.ff.net.2")
        key = key.replace(f"layers.{idx}.2.1", f"layers.{idx}.adaln_proj")
    image_proj_state_dict[key] = value
```

This is to account for the `IPAdapterTimeImageProjectionBlock` class that replaces ModuleList that previously was as:

```python
nn.ModuleList(
  [
      # msa
      TimePerceiverAttention(dim=hidden_dim, dim_head=dim_head, heads=heads),
      # ff
      nn.Sequential(
          nn.LayerNorm(hidden_dim),
          nn.Linear(hidden_dim, ff_inner_dim, bias=False),
          nn.GELU(),
          nn.Linear(ff_inner_dim, hidden_dim, bias=False),
      ),
      # adaLN
      nn.Sequential(
          nn.SiLU(),
          nn.Linear(hidden_dim, ff_inner_dim, bias=True)
      )
  ]
)
```

I think using existing `diffusers` components makes it easier to understand and minimizes the introduction of somewhat duplicated and single-use code, but does come with the downside of adjusting the state_dict. But as you said, there is only one checkpoint (afaik) available right now, so it's also the best time to make this decision.

I saw @yiyixuxu's message after making most of these changes, but it was a good exercise to learn more about diffusers nonetheless 😄 We can either roll back to the original implementation using `TimePerceiverAttention` or test the current version with `Attention` and `FusedAttnProcessor2_0`, as you prefer.

Reply:
I found that in fact these are not numerically equivalent, and one clear diference is how scale is applied in the dot product. In the PyTorch implementation, using [`F.scaled_dot_product_attention`](https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html#torch-nn-functional-scaled-dot-product-attention), we have:

```python
attn_weight = query @ key.transpose(-2, -1) * scale_factor
```

But in the `PerceiverAttention` module, we have:
```python
attn_weight = (query * scale_factor) @ (key * scale_factor).transpose(-2, -1)
```

Not sure why this difference exists, which one should we follow? If it's PyTorch's, we can use `FusedAttnProcessor2_0`. If we go with the `PerceiverAttention` implementation, we can just ignore the attention processor and compute it directly in `forward()` of `IPAdapterTimeImageProjectionBlock`.

I created a script to check output equality from both implementations and played around changing things, and in fact the only difference is the dot product op. I tried debugging with the PyTorch official equivalent implementation, but even that produces a different output from calling `F.scaled_dot_product_attention` (I found that somewhat weird, you can check the final part of the script).

<details>
<summary>Analysis script</summary>

```python
import math
import torch
import torch.nn as nn
import torch.nn.functional as F
from diffusers.models.embeddings import IPAdapterTimeImageProjection, IPAdapterTimeImageProjectionBlock, TimestepEmbedding, Timesteps
from examples.community.pipeline_stable_diffusion_3_ipa import TimeResampler, PerceiverAttention, reshape_tensor


DEVICE = "cuda"
DTYPE = torch.bfloat16

# ip_adapter from https://huggingface.co/InstantX/SD3.5-Large-IP-Adapter/resolve/main/ip-adapter.bin?download=true
state_dict = torch.load("ip-adapter.bin", "cpu")

image_proj_state_dict = {}
for key, value in state_dict["image_proj"].items():
    if key.startswith("layers."):
        idx = key.split(".")[1]
        key = key.replace(f"layers.{idx}.0.norm1", f"layers.{idx}.ln0")
        key = key.replace(f"layers.{idx}.0.norm2", f"layers.{idx}.ln1")
        key = key.replace(f"layers.{idx}.0.to_q", f"layers.{idx}.attn.to_q")
        key = key.replace(f"layers.{idx}.0.to_kv", f"layers.{idx}.attn.to_kv")
        key = key.replace(f"layers.{idx}.0.to_out", f"layers.{idx}.attn.to_out.0")
        key = key.replace(f"layers.{idx}.1.0", f"layers.{idx}.adaln_norm")
        key = key.replace(f"layers.{idx}.1.1", f"layers.{idx}.ff.net.0.proj")
        key = key.replace(f"layers.{idx}.1.3", f"layers.{idx}.ff.net.2")
        key = key.replace(f"layers.{idx}.2.1", f"layers.{idx}.adaln_proj")
    image_proj_state_dict[key] = value

# Image projetion parameters
embed_dim = image_proj_state_dict["proj_in.weight"].shape[1]
output_dim = image_proj_state_dict["proj_out.weight"].shape[0]
hidden_dim = image_proj_state_dict["proj_in.weight"].shape[0]
heads = image_proj_state_dict["layers.0.attn.to_q.weight"].shape[0] // 64
num_queries = image_proj_state_dict["latents"].shape[1]
timestep_in_dim = image_proj_state_dict["time_embedding.linear_1.weight"].shape[1]


############################ Full image proj test #############################
print("\n>> Image proj test")
# Image projection
ti_proj = IPAdapterTimeImageProjection(
    embed_dim=embed_dim,
    output_dim=output_dim,
    hidden_dim=hidden_dim,
    dim_head=64,
    heads=heads,
    num_queries=num_queries
).to(DEVICE, DTYPE)

tr = TimeResampler(
    dim=hidden_dim,
    depth=4,
    dim_head=64,
    heads=heads,
    num_queries=num_queries,
    embedding_dim=embed_dim,
    output_dim=output_dim
).to(DEVICE, DTYPE)

# Load weights
ti_proj.load_state_dict(image_proj_state_dict)
tr.load_state_dict(state_dict["image_proj"])

# Random variables
x = torch.rand(2, 279, 1152, device=DEVICE, dtype=DTYPE)
t = torch.rand(2, device=DEVICE, dtype=DTYPE)

latents_ti_proj, timestep_emb_ti_proj = ti_proj(x, t)
latents_tr, timestep_emb_tr = tr(x, t, need_temb=True)

print("image_proj out:", latents_tr.equal(latents_ti_proj))
print("timestep embed:", timestep_emb_tr.equal(timestep_emb_ti_proj))

###############################################################################


################################# Block test ##################################
print("\n>> Block attention test")

# Common projection params
time_proj = Timesteps(timestep_in_dim, True, 0).to(DEVICE, DTYPE)
time_embedding = TimestepEmbedding(timestep_in_dim, hidden_dim, act_fn="silu").to(DEVICE, DTYPE)
latents_params = nn.Parameter(image_proj_state_dict["latents"]).to(DEVICE, DTYPE)
proj_in = nn.Linear(embed_dim, hidden_dim).to(DEVICE, DTYPE)
proj_out = nn.Linear(hidden_dim, output_dim).to(DEVICE, DTYPE)
norm_out = nn.LayerNorm(output_dim).to(DEVICE, DTYPE)

# New implementation
block = IPAdapterTimeImageProjectionBlock(
    hidden_dim=hidden_dim,
    dim_head=64,
    heads=heads
).to(DEVICE, DTYPE)

# Original implementation
adaln_silu = nn.SiLU()
adaln_proj = nn.Linear(hidden_dim, 4 * hidden_dim, bias=True).to(DEVICE, DTYPE)
pa = PerceiverAttention(
    dim=hidden_dim,
    dim_head=64,
    heads=heads
).to(DEVICE, DTYPE)

# Load common params from state_dict
time_embedding.load_state_dict({
    "linear_1.weight": image_proj_state_dict["time_embedding.linear_1.weight"],
    "linear_1.bias": image_proj_state_dict["time_embedding.linear_1.bias"],
    "linear_2.weight": image_proj_state_dict["time_embedding.linear_2.weight"],
    "linear_2.bias": image_proj_state_dict["time_embedding.linear_2.bias"],
})
proj_in.load_state_dict({"weight": image_proj_state_dict["proj_in.weight"], "bias": image_proj_state_dict["proj_in.bias"]})
proj_out.load_state_dict({"weight": image_proj_state_dict["proj_out.weight"], "bias": image_proj_state_dict["proj_out.bias"]})
norm_out.load_state_dict({"weight": image_proj_state_dict["norm_out.weight"], "bias": image_proj_state_dict["norm_out.bias"]})

# Load first proj layer (original)
layer_0_pa_state_dict = {k.replace("layers.0.0.", ""): v for k,v in state_dict["image_proj"].items() if k.startswith("layers.0.0")}
pa.load_state_dict(layer_0_pa_state_dict)
adaln_proj.load_state_dict({"weight": state_dict["image_proj"]["layers.0.2.1.weight"], "bias": state_dict["image_proj"]["layers.0.2.1.bias"]})

# Load first proj layer (new)
layer_0_block_state_dict = {k.replace("layers.0.", ""): v for k,v in image_proj_state_dict.items() if k.startswith("layers.0.")}
block.load_state_dict(layer_0_block_state_dict)

# Random variables
x = torch.rand(2, 279, 1152, device=DEVICE, dtype=DTYPE)
t = torch.rand(2, device=DEVICE, dtype=DTYPE)

# Processing prior to attention
timestep_emb = time_proj(t).to(DEVICE, DTYPE)
timestep_emb = time_embedding(timestep_emb)

latents = latents_params.repeat(x.size(0), 1, 1)

x = proj_in(x)
x = x + timestep_emb[:, None]

x_input = x
latents_input = latents

############## Projection block - Original ##############
shift_msa1, scale_msa1, shift_mlp1, scale_mlp1 = adaln_proj(adaln_silu(timestep_emb)).chunk(4, dim=1)
        
x1_1 = pa.norm1(x_input)
latents1_1 = pa.norm2(latents_input)
latents1_2 = latents1_1 * (1 + scale_msa1.unsqueeze(1)) + shift_msa1.unsqueeze(1)

b1, l1, _ = latents1_2.shape

q1_1 = pa.to_q(latents1_2)
kv_input1 = torch.cat((x1_1, latents1_2), dim=-2)
k1_1, v1_1 = pa.to_kv(kv_input1).chunk(2, dim=-1)

q1_2 = reshape_tensor(q1_1, pa.heads)
k1_2 = reshape_tensor(k1_1, pa.heads)
v1_2 = reshape_tensor(v1_1, pa.heads)

scale1 = 1 / math.sqrt(math.sqrt(pa.dim_head))
weight1_1 = (q1_2 * scale1) @ (k1_2 * scale1).transpose(-2, -1) * scale1 # More stable with f16 than dividing afterwards
weight1_2 = torch.softmax(weight1_1.float(), dim=-1).type(weight1_1.dtype)
latents1_3 = weight1_2 @ v1_2

latents1_4 = latents1_3.permute(0, 2, 1, 3).reshape(b1, l1, -1)
latents1_5 = pa.to_out(latents1_4)

############## Projection block - New ##############
emb = block.adaln_proj(block.adaln_silu(timestep_emb))
shift_msa2, scale_msa2, shift_mlp2, scale_mlp2 = emb.chunk(4, dim=1)

x2_1 = block.ln0(x_input)
latents2_1 = block.ln1(latents_input) * (1 + scale_msa2[:, None]) + shift_msa2[:, None]
kv_input2 = torch.cat((x2_1, latents2_1), dim=-2)

b2, l2, _ = kv_input2.shape

q2_1 = block.attn.to_q(latents2_1)
kv2 = block.attn.to_kv(kv_input2)
split_size = kv2.shape[-1] // 2
k2_1, v2_1 = torch.split(kv2, split_size, dim=-1)

inner_dim = k2_1.shape[-1]
head_dim = inner_dim // block.attn.heads

q2_2 = q2_1.view(b2, -1, block.attn.heads, head_dim).transpose(1, 2)
k2_2 = k2_1.view(b2, -1, block.attn.heads, head_dim).transpose(1, 2)
v2_2 = v2_1.view(b2, -1, block.attn.heads, head_dim).transpose(1, 2)

# the output of sdp = (batch, num_heads, seq_len, head_dim)
latents2_2 = F.scaled_dot_product_attention(
    q2_2, k2_2, v2_2, attn_mask=None, dropout_p=0.0, is_causal=False, scale=block.attn.scale
)

latents2_3 = latents2_2.transpose(1, 2).reshape(b2, -1, block.attn.heads * head_dim)
latents2_4 = latents2_3.to(q2_2.dtype)

# linear proj
latents2_5 = block.attn.to_out[0](latents2_4)
# dropout
latents2_6 = block.attn.to_out[1](latents2_5)

latents2_7 = latents2_6 / block.attn.rescale_output_factor


############## Checks ##############
print("query:", q2_2.equal(q1_2))
print("key:", k2_2.equal(k1_2))
print("value:", v2_2.equal(v1_2))
print("sdpa out:", latents2_2.equal(latents1_3))
print("final output:", latents2_7.equal(latents1_5))

###############################################################################


################################# SDPA tests ##################################
print("\n>> SDPA tests")

# Efficient implementation equivalent to the following
# https://pytorch.org/docs/stable/generated/torch.nn.functional.scaled_dot_product_attention.html
def eq_scaled_dot_product_attention(query, key, value, attn_mask=None, dropout_p=0.0,
        is_causal=False, scale=None, enable_gqa=False) -> torch.Tensor:
    L, S = query.size(-2), key.size(-2)
    scale_factor = 1 / math.sqrt(query.size(-1)) if scale is None else scale
    attn_bias = torch.zeros(L, S, dtype=query.dtype, device=query.device)
    if is_causal:
        assert attn_mask is None
        temp_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0)
        attn_bias.masked_fill_(temp_mask.logical_not(), float("-inf"))
        attn_bias.to(query.dtype)

    if attn_mask is not None:
        if attn_mask.dtype == torch.bool:
            attn_bias.masked_fill_(attn_mask.logical_not(), float("-inf"))
        else:
            attn_bias += attn_mask

    if enable_gqa:
        key = key.repeat_interleave(query.size(-3)//key.size(-3), -3)
        value = value.repeat_interleave(query.size(-3)//value.size(-3), -3)

    attn_weight = query @ key.transpose(-2, -1) * scale_factor
    attn_weight += attn_bias
    attn_weight = torch.softmax(attn_weight, dim=-1)
    attn_weight = torch.dropout(attn_weight, dropout_p, train=True)
    return attn_weight @ value

for backend in (
    nn.attention.SDPBackend.FLASH_ATTENTION,
    nn.attention.SDPBackend.EFFICIENT_ATTENTION,
    nn.attention.SDPBackend.MATH
):
    with nn.attention.sdpa_kernel(backend):
        # This should be equal?
        print(
            f"Equality check with backed {backend}:",
            F.scaled_dot_product_attention(
                q2_2, k2_2, v2_2, attn_mask=None, dropout_p=0.0, is_causal=False, scale=1.0
            ).equal(
                eq_scaled_dot_product_attention(
                    q2_2, k2_2, v2_2, attn_mask=None, dropout_p=0.0, is_causal=False, scale=1.0
                )
            )
        )
``` 
</details>

The script outputs the following:
```
>> Image proj test
image_proj out: False
timestep embed: True

>> Block attention test
query: True
key: True
value: True
sdpa out: False
final output: False

>> SDPA tests
Equality check with backed SDPBackend.FLASH_ATTENTION: False
Equality check with backed SDPBackend.EFFICIENT_ATTENTION: False
Equality check with backed SDPBackend.MATH: False
```

Reply:
It's ok to just use TimePerceiverAttention

Reply:
Done, now the outputs are numerically equivalent

```python
import torch
from diffusers.models.embeddings import IPAdapterTimeImageProjection
from examples.community.pipeline_stable_diffusion_3_ipa import TimeResampler


DEVICE = "cuda"
DTYPE = torch.bfloat16

# ip_adapter from https://huggingface.co/InstantX/SD3.5-Large-IP-Adapter/resolve/main/ip-adapter.bin?download=true
state_dict = torch.load("ip-adapter.bin", "cpu", weights_only=True)

# Convert image_proj state dict to diffusers
image_proj_state_dict = {}
for key, value in state_dict["image_proj"].items():
    if key.startswith("layers."):
        idx = key.split(".")[1]
        key = key.replace(f"layers.{idx}.0.norm1", f"layers.{idx}.ln0")
        key = key.replace(f"layers.{idx}.0.norm2", f"layers.{idx}.ln1")
        key = key.replace(f"layers.{idx}.0.to_q", f"layers.{idx}.attn.to_q")
        key = key.replace(f"layers.{idx}.0.to_kv", f"layers.{idx}.attn.to_kv")
        key = key.replace(f"layers.{idx}.0.to_out", f"layers.{idx}.attn.to_out.0")
        key = key.replace(f"layers.{idx}.1.0", f"layers.{idx}.adaln_norm")
        key = key.replace(f"layers.{idx}.1.1", f"layers.{idx}.ff.net.0.proj")
        key = key.replace(f"layers.{idx}.1.3", f"layers.{idx}.ff.net.2")
        key = key.replace(f"layers.{idx}.2.1", f"layers.{idx}.adaln_proj")
    image_proj_state_dict[key] = value

# Image projetion parameters
embed_dim = image_proj_state_dict["proj_in.weight"].shape[1]
output_dim = image_proj_state_dict["proj_out.weight"].shape[0]
hidden_dim = image_proj_state_dict["proj_in.weight"].shape[0]
heads = image_proj_state_dict["layers.0.attn.to_q.weight"].shape[0] // 64
num_queries = image_proj_state_dict["latents"].shape[1]
timestep_in_dim = image_proj_state_dict["time_embedding.linear_1.weight"].shape[1]

# Image projection
ti_proj = IPAdapterTimeImageProjection(
    embed_dim=embed_dim,
    output_dim=output_dim,
    hidden_dim=hidden_dim,
    dim_head=64,
    heads=heads,
    num_queries=num_queries
).to(DEVICE, DTYPE)

tr = TimeResampler(
    dim=hidden_dim,
    depth=4,
    dim_head=64,
    heads=heads,
    num_queries=num_queries,
    embedding_dim=embed_dim,
    output_dim=output_dim
).to(DEVICE, DTYPE)

# Load weights
ti_proj.load_state_dict(image_proj_state_dict)
tr.load_state_dict(state_dict["image_proj"])

# Random variables
x = torch.rand(2, 279, 1152, device=DEVICE, dtype=DTYPE)
t = torch.rand(2, device=DEVICE, dtype=DTYPE)

latents_ti_proj, timestep_emb_ti_proj = ti_proj(x, t)
latents_tr, timestep_emb_tr = tr(x, t, need_temb=True)

print("image_proj out:", latents_tr.equal(latents_ti_proj))
print("timestep embed:", timestep_emb_tr.equal(timestep_emb_ti_proj))
```